File: src/common/web/data/entities/project/features/MetadataFeature.ts
import { ProjectObject } from "@common/ui/components/propertyeditor/ProjectObjectStore";
import { ProjectFeature, type ProjectFeatureID } from "./ProjectFeature";

/**
 * The project metadata type.
 *
 * TODO: Use proper type
 */
export type ProjectMetadata = ProjectObject[];

/**
 * The metadata objects type.
 *
 * TODO: Use proper type
 */
export type MetadataObjects = ProjectObject[];

/**
 * Data class for the metadata project feature.
 */
export class MetadataFeature extends ProjectFeature {
    public static readonly FeatureID: ProjectFeatureID = "metadata";

    public readonly metadata: ProjectMetadata;
    public readonly shared_objects: MetadataObjects;

    public constructor(metadata: ProjectMetadata = [], sharedObjects: MetadataObjects = []) {
        super();

        this.metadata = metadata;
        this.shared_objects = sharedObjects;
    }

    public get featureID(): ProjectFeatureID {
        return MetadataFeature.FeatureID;
    }
}

File: src/common/py/data/entities/project/features/metadata_feature.py
import typing
from dataclasses import dataclass, field

from dataclasses_json import dataclass_json

from .project_feature import ProjectFeature, ProjectFeatureID

ProjectMetadata = typing.List[typing.Dict[str, typing.Any]]  # TODO: Use proper type


@dataclass_json
@dataclass
class MetadataFeature(ProjectFeature):
    """
    Data class for the metadata project feature.
    """

    feature_id: typing.ClassVar[ProjectFeatureID] = "metadata"

    metadata: ProjectMetadata = field(default_factory=list)
